protobuf-format.md

```diff
@@ -0,0 +1,227 @@
+
+This approach allows for attributes to be represented and transported
+with no loss of *type* information.
+
+## 3. Data
+
+The specification allows for data payloads of the following types to be explicitly represented:
+
+* string
+* bytes
+* protobuf object/message
+
+```proto
+oneof data_oneof {
+
+    // Binary data
+    bytes binary_data = 2;
+
+    // String data
+    string text_data = 3;
+
+    // Protobuf Message data
+    google.protobuf.Any proto_data = 4;
+}
+```
+
+* Where the data is a protobuf message it MUST be stored in the `proto_data` property.
+  * `datacontenttype` MAY be populated with `application/protobuf`
+  * `dataschema` SHOULD be populated with the type URL of the protobuf data message.
+
+* When the type of the data is text, the value MUST be stored in the `text_data` property.
+  * `datacontenttype` SHOULD be popuplated with the appropriate media-type.
+
+* When the type of the data is binary the value MUST be stored in the `binary_data` property.
+  * `datacontenttype` SHOULD be populated with the appropriate media-type.
+
+
+
+## 4. Transport
+
+Transports that support content identification MUST use the following designation:
+
+```text
+   application/cloudevents+protobuf
+```
+
+## 5. Examples
+
+The following code-snippets shows how proto representations might be constucted asuming the availability of some convenience methods ...
+
+### 5.1 Plain Text event data
+
+```java
+public static Spec.CloudEvent plainTextExample() {
+  Spec.CloudEvent.Builder ceBuilder = Spec.CloudEvent.newBuilder();
+
+  ceBuilder
+    //-- Required Attributes.
+    .setId(UUID.randomUUID().toString())
+    .setSpecVersion("1.0")
+    .setType("io.cloudevent.example")
+    .setSource("producer-1")
+
+    //-- Data.
+    .setTextData("This is a plain text message");
+
+    //-- Optional Attributes
+    withCurrentTime(ceBuilder, "time");
+    withAttribute(ceBuilder, "datacontenttype", "text/plain");
+
+    // Build it.
+    return ceBuilder.build();
+}
+
+```
+
+### 5.2 Proto message as event data
+
+Where the event data payload is itself a protobuf message (with its own schema)
+a protocol buffer idiomatic method can be used to carry the data.
+
+```java
+private static Spec.CloudEvent protoExample() {
+
+  //-- Build an event data protobuf object.
+  Test.SomeData.Builder dataBuilder = Test.SomeData.newBuilder();
+
+  dataBuilder
+    .setSomeText("this is an important message")
+    .setIsImportant(true);
+
+  //-- Build the CloudEvent.
+  Spec.CloudEvent.Builder ceBuilder = Spec.CloudEvent.newBuilder();
+
+  ceBuilder
+    .setId(UUID.randomUUID().toString())
+    .setSpecVersion("1.0")
+    .setType("io.cloudevent.example")
+    .setSource("producer-2")
+
+    // Add the proto data into the CloudEvent envelope.
+    .setProtoData(Any.pack(dataBuilder.build()));
+
+    // Add the protto type URL
+    withAttribute(ceBuilder, "dataschema", ceBuilder.getProtoData().getTypeUrl());
```

Comment:
I'm not keen on `with` as a name - I'd prefer `set`. I generally expect a `with` method to return a new instance with the specified change applied, leaving the original object as it was before.